scormxblock/scorm_file_uploader.py

```diff
@@ -0,0 +1,166 @@
+        self.temp_file_path = os.path.join(tempfile.gettempdir(), xblock.location.block_id)
+        self.scorm_storage_location = os.path.join(scorm_storage_location, xblock.location.block_id)
+
+    def upload(self):
+        content_range = self._get_content_range()
+
+        if int(content_range['start']) == 0:
+            mode = "wb+"
+        else:
+            mode = "ab+"
+            size = os.path.getsize(self.temp_file_path)
+            if size > int(content_range['stop']) and size == int(content_range['end']):
+                return STATES.complete, None
+
+        self._write_to_file(mode=mode)
+        size = os.path.getsize(self.temp_file_path)
+
+        if int(content_range['stop']) != int(content_range['end']) - 1:
+            # More chunks coming
+            return STATES.progress, size
+
+        scorm_file_url = self._extract_and_store()
+
+        return STATES.complete, scorm_file_url
+
+    def _extract_and_store(self):
+        cache.set('upload_percent_{}'.format(self.xblock.location.block_id), 0, 1 * 60 * 60)
+        unizpped_dir = self._extract_zipped_file()
+        storage_url = self._save_to_storage(unizpped_dir)
+
+        return storage_url
+
+    def _save_to_storage(self, tempdir):
+        storage = self._get_storage()
+        self._cleanup_storage_dir(storage)
+
+        to_store = []
+        total_size = 0
+        for (dirpath, dirnames, files) in os.walk(tempdir):
+            for f in files:
+                file_path = os.path.join(os.path.abspath(dirpath), f)
+                size = os.path.getsize(file_path)
+                total_size += size
+                to_store.append({'path': file_path, 'size': size})
+
+        uploaded_size = 0
+        for f in to_store:
+            # defensive decode/encode from zip
+            file_path = f['path']
+            f_path = file_path.decode(self.xblock.encoding).encode('utf-8').replace(tempdir, '')
+            with open(file_path, 'rb+') as fh:
+                try:
+                    logger.info('Storing file `{}` of size `{}` on S3'.format(f_path, f['size']))
+                    storage.save('{}{}'.format(self.scorm_storage_location, f_path), fh)
+                    logger.info('File `{}` stored.'.format(f_path))
+                    uploaded_size += f['size']
+                    self._set_upload_progress(uploaded_size, total_size)
+                except encoding.DjangoUnicodeDecodeError, e:
+                    logger.warn('SCORM XBlock Couldn\'t store file {} to storage. {}'.format(f, e))
+
+        self._post_upload_cleanup(tempdir)
+
+        url = storage.url(self.scorm_storage_location)
+        return '?' in url and url[:url.find('?')] or url
+
+    def _set_upload_progress(self, uploaded, total):
+        percent = int(uploaded / float(total) * 100)
+        block_id = self.xblock.location.block_id
+
+        cache.set('upload_percent_{}'.format(block_id), percent, 1 * 60 * 60)
```

Comment:
Are we sure no package can take longer than 1hr to upload completely?

Reply:
This key is set (i.e; progress updated) each time a single file completes upload. Packages can take longer or less, this time ensures that during upload of a single file, cache don't get expired. So the assumption is that a single file won't take longer than an hour